refactor(core): optimize the implementation about finding context from error in ErrorHandler

damingerdai opened the pull request:
## Changes

in _findContext method, first check whether the params 'error' exists and then use '||' instead conditional operator when getDebugContext's result does not exist.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Commit:
refactor(core): optimize the implementation about finding context from error in ErrorHandler

in _findContext method, use conditional operator  check whether the params 'error' exists and then use nullish coalescing operator instead conditional operator when getDebugContext's result does not exist.

dario-piotrowicz commented on packages/core/src/error_handler.ts:
```diff
@@ -59,12 +59,11 @@ export class ErrorHandler {
     }
 
-    return null;
+    return getDebugContext(error) || this._findContext(getOriginalError(error));
```
If we wanted we could even avoid the `return null` and make this a one liner (not sure if it's less readable)

Also I think that maybe the nullish coalescing operator here would be more appropriate than the or, wouldn't it?

```suggestion
    return !error ? null : getDebugContext(error) ?? this._findContext(getOriginalError(error));
```

damingerdai replied:
thanks for your suggestion. please review again.

dario-piotrowicz reviewed (approved):
Cool, thanks for the update :)

AndrewKushnir commented on packages/core/src/error_handler.ts:
```diff
@@ -59,12 +59,7 @@ export class ErrorHandler {
-    }
-
-    return null;
+    return !error ? null : getDebugContext(error) ?? this._findContext(getOriginalError(error));
```
Since nullish coalescing is a new operator, which is not supported natively in many browsers, TypeScript will downlevel it (change it to make compatible with a given target) while emitting JS for the targets older than ES2020. The code that you've updated is a part of Angular runtime, so it'd be included into app bundles (including prod ones I believe), so it'd be great to keep it as small as possible. I've put together [a quick example using TypeScript playground](https://www.typescriptlang.org/play?target=2#code/GYVwdgxgLglg9mABAcwKZQCKoEYmQYQSlQA8oAKVAJyrioC5EBDMATwEpGXXEBvAXwCwAKFCRYCFOgDyVGMhhgmAGwCiNOpQ0NmbdnyHCREZUwDOZxOtpUAEiwAmy6nxGJEAfWCKHhMMTItGy49AxE3RAB6SMQAZThlEAkkAGIARgA6RHwQGlR-RAA3ajN4MAyIrx8-AKgPCFyqfIpqYN0OEJ5eCPcYYEQgun1u4XcxsaaoXKQ0TBw8GtIW7X0Afik53AIiJcGqfXoe8eOT07PTqAALGDMMqrBfHcDZ2XlFFWtNVqH2AG4jwzjSbTRBgEDKZT-UZjSIAKkQAFlFH0YKgHEUSmUBgA2AAMiAglyYVDMnCO4z6lHYvGBVBmVNWyCp9CuNzu3gei0CcCpf34tKQYIhUPGsMiEUMEWicQSSSxKQATFkAJIAW2UAAdUKrmkxkogNbQNXAzGjEIpEFdUIhgAllHAAO6KZCIAAKACVEJcoFANWZ6NEFFcQNgMhA4KrIixkODiVGwDHTFRIhrwcpIgAWBUAVgAHGlKhzHv4ll5aKqPB69p1OO1XND3ALEN8qIh1uRZlgtlzljY1utWbd7sXah2ZHIFEo1No9uw54hGELIUc4Yjkd4zcUSVjyLn8YTiaTDg3joViZ4mL8m6hVuRyB4mABeJmoOcAQkfj6XADJvw+P4+hRwDA6K4qsD4stcQ5Fj25A8q+c70EuIpjGKErhNC0rxIk+opAAzFkrpGiaKg2uWbqek0hSog6AA0iCmta3q+v6gYwMGobhpG0axsmPFJimaaZjm+aFtUTx1FRNHVu0tbcPWQLoCCLZtgMnbzNsJaBC2+gAD66ZaUHsuJWkUC8E7vNONizvOi5pih7irkiYAopumKSOQGbYgSRIkmSJ6KVMdLNreL7sPpg7GZyElwbynDIUcaHQoCIj8EAA) to illustrate different options, how they look like in JS code and how they may look like in minified version. Based on my investigation, it looks like the following solution would be the most optimal one (good readability + no downleveling overhead):

```suggestion
    return error ? (getDebugContext(error) || this._findContext(getOriginalError(error))) : null;
```
Note: I've inverted the `!error` condition and also added parenthesis to wrap sub-expression for readability.

What do you think?

Just also wanted to mention that we use nullish coalescing and other modern syntax in various parts of the codebase where having extra bytes is not critical (for example in compiler's code which we do not ship to the browser, except for JIT use-case, which is not that common). However in runtime code we try to find the right balance and avoid extra payload size where possible (**without** sacrificing readability, code quality, etc).

dario-piotrowicz replied:
yeah I suggested the coalescing operator as it seemed more "safe" to me, in that way no assumption needs to be made on the `getDebugContext` result (although it is quite unlikely to get something falsy from that function regardless anyways I think)

so yeah I do believe that you are making a very valid point :slightly_smiling_face: 

Although the js output difference is very small and this code isn't going to run too often (as it is error handling not something that should happen all the time right?) there is indeed a difference and I think as you said given that this is runtime it should indeed be kept as small as possible, so I totally agree with you

Thank you so very much for looking into that, your comment is really great :slightly_smiling_face: 

> note: not sure about the `!error` adding parenthesis, you may be right, but to me it looked a bit cleaner to get the short branch first otherwise you'd have to read the whole thing to understand what's going on, also I know that some people don't like ternaries with negated conditions so that's also a minus anyhow :P

AndrewKushnir replied:
Thanks for the reply and for sharing your feedback 👍

jessicajaniuk commented:
There seem to be 4 consistently failing tests in the `test_ivy_aot` test run related to golden files. You may need to rebase / take a look at the failures.

damingerdai commented:
> There seem to be 4 consistently failing tests in the `test_ivy_aot` test run related to golden files. You may need to rebase / take a look at the failures.

i have rebased the last master branch.

Commit:
build: update all symbol extractor

update all symbol extractor by running 'yarn symbol-extractor:update' script

AndrewKushnir commented:
Hi @damingerdai, thanks for improving the code. Could you please have a look at [my comment](https://github.com/angular/angular/pull/42581/files#r653121007) in this thread and update the code to use `||` (vs `??`)? Thank you.

Commit:
refactor(core): optimize the implementation about finding context from error in ErrorHandler

in _findContext method, use conditional operator check whether the params 'error' exists and then us reccursion way to find context in original error  if getDebugContext's result does not exist.

damingerdai commented:
> Hi @damingerdai, thanks for improving the code. Could you please have a look at [my comment](https://github.com/angular/angular/pull/42581/files#r653121007) in this thread and update the code to use `||` (vs `??`)? Thank you.
thanks for [your review](https://github.com/angular/angular/pull/42581/files#r653121007), i have update my code.

AndrewKushnir reviewed (approved):
@damingerdai thanks for addressing the feedback. The change looks good 👍 

I've started tests in Google's codebase ([internal-only link](https://test.corp.google.com/ui#id=OCL:381160823:BASE:384266945:1626115559715:43d18087)) and if everything goes well, we'll add this PR to the merge queue.

AndrewKushnir commented:
@damingerdai thanks for the optimization performed in this PR! 👍 
The change is now merged and will be available on NPM after the upcoming Angular release this week.
